File: Week_01/G20200343030400/LeetCode_22_400.cpp
//22
//https://leetcode-cn.com/problems/generate-parentheses

class Solution {
public:
    void recursion(string temp, int left_possible, int right_possible, vector<string>& ans) {
        if (left_possible == 0 && right_possible == 0)
            ans.push_back(temp);
        if (left_possible > 0)
            recursion(temp + "(", left_possible - 1, right_possible + 1, ans);
        if (right_possible >0) {
            recursion(temp + ")", left_possible, right_possible - 1, ans);
        }
    }

    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        recursion("", n, 0, ans);
        return ans;
    }
};

File: Week_01/G20200343030400/LeetCode_22_400.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if n == 0:
            return []
        
        ans = []
        
        def recursion(left_possible, right_possible, temp, ans):
            if left_possible == right_possible == 0:
                ans.append(temp)
            if left_possible > 0:
                recursion(left_possible - 1, right_possible + 1, temp + '(', ans)
            if right_possible > 0:
                recursion(left_possible, right_possible - 1, temp + ')', ans)
        
        recursion(n, 0, '', ans)
        return ans
